code/artificial_intelligence/src/a_star/astar.c

```diff
@@ -0,0 +1,253 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <float.h>
+#include <iso646.h>
+#include <math.h>
+ 
+#define map_size_rows 10
+#define map_size_cols 10
+ 
+char map[map_size_rows][map_size_cols] = {
+    {1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
+    {1, 0, 0, 0, 0, 0, 0, 0, 0, 1},
+    {1, 0, 0, 0, 0, 0, 0, 0, 0, 1},
+    {1, 0, 0, 0, 0, 1, 1, 1, 0, 1},
+    {1, 0, 0, 1, 0, 0, 0, 1, 0, 1},
+    {1, 0, 0, 1, 0, 0, 0, 1, 0, 1},
+    {1, 0, 0, 1, 1, 1, 1, 1, 0, 1},
+    {1, 0, 0, 0, 0, 0, 0, 0, 0, 1},
+    {1, 0, 0, 0, 0, 0, 0, 0, 0, 1},
+    {1, 1, 1, 1, 1, 1, 1, 1, 1, 1}
+};
+ 
+// description of graph node
+struct stop {
+    double col, row;
+    // array of indexes of routes from this stop to neighbours in array of all routes
+    int *n;
+    int n_len;
+    double f, g, h;
+    int from;
+};
+ 
+int ind[map_size_rows][map_size_cols] = {
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1},
+    {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1}
+};
+ 
+// description of route between two nodes
+struct route {
+    // route has only one direction!
+    int x; // index of stop in array of all stops of src of this route
+    int y; // intex of stop in array of all stops od dst of this route
+    double d;
+};
+ 
+int 
+main() {
+    int i, j, k, l, b, found;
```

Comment:
All loop variables should be declare as close to the loop as possible. All other variables (including those below) should be declared where they are used.